tests/test_pickling.py

```diff
@@ -1,25 +1,47 @@
 import copy
 import pickle
-import re
 
 import pytest
 
 import env
 from pybind11_tests import pickling as m
 
 
-def test_pickle_simple_callable():
+def test_assumptions():
+    assert pickle.HIGHEST_PROTOCOL >= 0
+
+
+@pytest.mark.parametrize("protocol", range(pickle.HIGHEST_PROTOCOL + 1))
+def test_pickle_simple_callable(protocol):
     assert m.simple_callable() == 20220426
-    if env.PYPY:
-        serialized = pickle.dumps(m.simple_callable)
-        deserialized = pickle.loads(serialized)
-        assert deserialized() == 20220426
-    else:
-        # To document broken behavior: currently it fails universally with
-        # all C Python versions.
-        with pytest.raises(TypeError) as excinfo:
-            pickle.dumps(m.simple_callable)
-        assert re.search("can.*t pickle .*PyCapsule.* object", str(excinfo.value))
+    serialized = pickle.dumps(m.simple_callable, protocol=protocol)
+    assert b"pybind11_tests.pickling" in serialized
+    assert b"simple_callable" in serialized
+    deserialized = pickle.loads(serialized)
+    assert deserialized() == 20220426
+    assert deserialized is m.simple_callable
+
+    # UNUSUAL: A pickle roundtrip starting with `m.simple_callable.__self__` yields `m`:
```

Comment:
Let's hit a few keywords to hopefully make it more likely to land here from a google search:

# UNUSUAL: function record pickling roundtrip returns a module, not the function record object:

Reply:
Done: 1c24995d75c2b54443191d8bcba2f5d927022cdd

Thanks!